### PR TITLE
More length support

### PR DESCRIPTION
**Pull request by shans**

### CODE REVIEW

File: web-animation.js [hunk 1]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
+var outerCalcRE = /-webkit-calc\s*\(\s*([^)]*)\)/
+var valueRE = /\s*([0-9.]*)([a-zA-Z%]*)/
+var operatorRE = /\s*([+-])/
+
 var percentLengthType = {
-  zero: function() { return {px: 0, percent: 0}; },
+  zero: function() { return {}; },
   add: function(base, delta) { 
-    return {px: base.px + delta.px, 
-        percent: base.percent + delta.percent};
```
Comment (dstoc) [inline web-animation.js]: Does this need min/max?

File: web-animation.js [hunk 2]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
+    out = {}
```
Comment (dstoc) [inline web-animation.js]: var out

File: web-animation.js [hunk 3]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
+            continue;
+        }
+        out[value] = delta[value];
+    }
+    return out;
   },
   interpolate: function(from, to, f) {
-    return {px: interp(from.px, to.px, f), 
-        percent: interp(from.percent,to.percent, f)};
+    out = {}
```
Comment (dstoc) [inline web-animation.js]: var out

File: web-animation.js [hunk 4]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
+    for (value in from) {
```
Comment (DmitryBaranovskiy) [inline web-animation.js]: var value

File: web-animation.js [hunk 5]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
+var operatorRE = /\s*([+-])/
+
 var percentLengthType = {
-  zero: function() { return {px: 0, percent: 0}; },
+  zero: function() { return {}; },
   add: function(base, delta) { 
-    return {px: base.px + delta.px, 
-        percent: base.percent + delta.percent};
+    out = {}
+    for (value in base) {
```
Comment (DmitryBaranovskiy) [inline web-animation.js]: var value

File: web-animation.js [hunk 6]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
 
 var calcRE = /-webkit-calc\s*\(\s*([^+\s)]*)\s*([+-])\s*([^+\s)]*)\s*\)/
 
+var outerCalcRE = /-webkit-calc\s*\(\s*([^)]*)\)/
+var valueRE = /\s*([0-9.]*)([a-zA-Z%]*)/
+var operatorRE = /\s*([+-])/
```
Comment (DmitryBaranovskiy) [inline web-animation.js]: semicolons for above three lines, please

File: web-animation.js [hunk 7]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
-    return {px: base.px + delta.px, 
-        percent: base.percent + delta.percent};
+    out = {}
+    for (value in base) {
+        if (value in delta) {
+            out[value] = base[value] + delta[value];
+        } else {
+            out[value] = base[value];
+        }
+    }
```
Comment (DmitryBaranovskiy) [inline web-animation.js]: instead  of above `if` you could write

```
out[value] = base[value] + (delta[value] || 0);
```

(just a suggestion)

File: web-animation.js [hunk 8]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
   toCssValue: function(value) { 
-    if (value.percent == 0) {
-      return value.px + 'px';
-    } else if (value.px == 0) {
-      return value.percent + '%';
-    } else {
-      return '-webkit-calc(' + value.px + 'px + ' + value.percent + '%)';
+    var s = '';
+    single_value = true;
+    for (item in value) {
```
Comment (DmitryBaranovskiy) [inline web-animation.js]: `var single_value`
`var item`

File: web-animation.js [hunk 9]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
+    }
+    innards = innards[1];
+    var first_time = true;
+    while (true) {
+      var reversed = false;
+      if (first_time) {
+        first_time = false;
+      } else {
+        var op = operatorRE.exec(innards);
+        if (!op) {
```
Comment (DmitryBaranovskiy) [inline web-animation.js:1289]: Inconsistency with line 1539. Either `if (!op)` or `if (op == null)`. I would vote for the first case.

File: web-animation.js [hunk 10]
```diff
@@ -1482,41 +1482,100 @@ var numberType = {
...
+      }
+      value = valueRE.exec(innards);
+      if (!value) {
+        return {};
+      }
+      if (!isDefinedAndNotNull(out[value[2]])) {
+        out[value[2]] = 0;
+      }
+      if (reversed) {
+        out[value[2]] -= Number(value[1]);
```
Comment (DmitryBaranovskiy) [inline web-animation.js:1305]: Superfluous conversion to number.

File: web-animation.js (hunk 1 above)
Comment (shans) [inline web-animation.js]: I don't think so? Many values take negative lengths.

File: web-animation.js (hunk 10 above)
Comment (shans) [inline web-animation.js:1305]: I prefer an explicit conversion for clarity.

**Comment by shans** [comment]
responded to review comments and added a test.

File: web-animation.js [hunk 11]
```diff
@@ -1475,43 +1475,96 @@ var integerType = {
...
 var percentLengthType = {
-  zero: function() { return {px: 0, percent: 0}; },
-  add: function(base, delta) {
-    return {px: base.px + delta.px,
-        percent: base.percent + delta.percent};
+  zero: function() { return {}; },
+  add: function(base, delta) { 
+    var out = {};
+    for (value in base) {
+        out[value] = base[value] + (delta[value] || 0);
```
Comment (dstoc) [inline web-animation.js]: indentation is off

File: web-animation.js [hunk 12]
```diff
@@ -1475,43 +1475,96 @@ var integerType = {
...
-      if (r) {
-        var left = lengthType.fromCssValue(r[1]);
-        var right = lengthType.fromCssValue(r[3]);
-        if (r[2] == '-') {
-          right = {px: -right.px, percent: -right.percent};
+    var out = {}
+    var innards = outerCalcRE.exec(value);
+    if (!innards) {
+      var singleValue = valueRE.exec(value);
+      if (singleValue.length == 3) {
```
Comment (dstoc) [inline web-animation.js]: if (singleValue)

**Comment by shans** [comment]
Latest comments applied.

**Pull request merged**